[FIRRTL][HW] Remove dead named wires

uenoku opened the pull request:
This PR removes dead wires created by name preservation of FIRRTL based on https://github.com/llvm/circt/pull/2851.
1. At LowerToHW, "user_specified_name" attribute is attached to wires created by name preservation nodes/wires. 
2. At Prettify, we remove dead wires with "user_specified_name" attribute. 

I'm afraid that removing named wires at Prettify is abusing PrettifyVerilog pass.

mwachs5 commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -2341,14 +2341,21 @@ LogicalResult FIRRTLLowering::visitDecl(WireOp op) {
         Twine("__") + moduleName + Twine("__") + name.getValue()));
   }
+
+  bool hasChiselName = false;
```
perhaps instead of `ChiselName` could be something like `hasUserSpecifiedName` ?

uenoku replied:
Thanks, I agree! I change to `userSpecifiedName`.

mwachs5 commented on test/firtool/name-preservation.fir:
```diff
@@ -13,9 +13,9 @@ circuit Foo:
-    ; Default behavior is to preserve named wires.
-    ; CHECK:        wire x_a;
-    ; CHECK:        wire x_b;
+    ; Default behavior is to preserve named wires unless they are not dead
```
```suggestion
    ; Default behavior is to preserve named wires unless they are dead
```

...?

uenoku replied:
Yes, thanks!

mikeurbach reviewed (commented):
Thanks for this change, it will clean up a ton of lint warnings.

Regarding some of the implementation details: I had thought about doing this at Prettify originally, and it sort of makes sense to me there. Another thought I had is to modify the sv.wire canonicalizer to not exit early if the wire has a symbol, but the "user_specified_name" attribute is present. Would that achieve the same behavior? https://github.com/llvm/circt/blob/72838a205cb11846743bd0724493eafbaf65cc5c/lib/Dialect/SV/SVOps.cpp#L1166-L1168

Also- I think stamping out a unit attribute could be a good way to implement this. Should it be a dialect attribute in the firrtl dialect? I think it is proper form to use the dialect prefix on such discardable attributes. I'm thinking something like these (and documenting this one there as well): https://circt.llvm.org/docs/FIRRTLAnnotations/#attributes-in-sv

mikeurbach commented:
OTOH, doing FIRRTL-dialect specific things in the canonicalizer for sv.wire feels icky. Maybe this should be a dialect attribute in the SV dialect (whether this change occurs in prettify, canonicalizer, or elsewhere).

mikeurbach commented:
> With allowing the canonicaliztion:
> 
>   assign b = a; // foo.fir:2:10

I think some have actually argued for this, but let's probably avoid that for now. Good to know if we go down that rabbit hole we can implement it in the canonicalizer.

> I guess "sv.namehint" has different semantics from "user_specified_name" in terms of preservation.

Yeah that makes sense. I guess my request is just to modify "user_specified_name" -> "sv.user_specified_name", and document its behavior.

darthscsi commented on docs/RationaleSV.md:
no.  This defeats the purpose of a symbol.  If you don't need the wire to appear in the output, strip the symbol.

uenoku replied:
It makes sense. Note that it is necessary to switch the behavior of several SV canonicalizers based on the appearance of the attribute if we don't use a symbol.

mikeurbach replied:
Could this be handled then in LowerToHW, when it's deciding to add a symbol for name preservation? Instead of adding this unit attr, move the check currently in Prettify into LowerToHW and gate adding a symbol based on that?

uenoku replied:
That's a good point. Probably FIRRTL canonicalizer can delete dead wires/nodes created by name preservations. If doing so, it would not be required to change LowerToHW (because if named wires/nodes reach to LowerToHW, it means they are alive).

mikeurbach replied:
If done at the canonicalizer, would that lead to this situation?

> With allowing the canonicaliztion:
>
>  assign b = a; // foo.fir:2:10

I'm not sure if that would happen, but I was thinking doing it at LowerToHW might avoid this.

There's also an argument that we should be doing that anyway...

uenoku replied:
I think we are currently not named nodes at FIRRTL even when nodes are dead because it was expected behavior of name preservation. However, if it is allowed to eliminate dead named nodes, I guess we can add that canonicalization.

mikeurbach replied:
I see, well if it can be done as a FIRRTL canonicalization, I think that makes sense.

uenoku commented:
I added FIRRTL caononicalization to erase dead named wires. However, it was not sufficient because comb canonicalizers create dead wires. Therefore, anyway it is necessary to eliminate dead wires at HW/SV level as well. I moved the elimination logic from Prettify to SV canonicalizer, which I think more appropriate place for the canonicalizer to live.

mikeurbach commented:
Looking at @darthscsi's previous comment, and the now documented rationale (https://github.com/llvm/circt/blob/main/docs/RationaleFIRRTL.md#symbols-and-inner-symbols), I think the current implementation might still be controversial. Is it possible to change LowerToHW to not assign a symbol if the wire/node is not read from? I think this is the gist of what @darthscsi meant before.

uenoku commented:
The new implementation doesn't create symbols. Instead, `sv.user_specified_name` prevents optimizations.

uenoku commented:
I'll change the PR following @darthscsi's suggestion in the following way:
1. Remove SV changes. SV wire doesn't have a mechanism to model FIRRTL's interesting/useless name. FIRRTL canonicalizers should be strong enough to remove dead named wires.
2. Move FIRRTL canonicalizer changes into the dedicated pass. We have to preserve the naming semantics of chisel at FIRRTL canonicalizers. Instead, we should create "DropDeadInterestingNamePass" and invoke the pass only when optimization level is >= O1.

uenoku commented:
We will use different mechanism so close this.
